Blog: Scaling OpenTelemetry Collectors using Ansible (#4182)
open-telemetry/opentelemetry.io: content/en/blog/2024/scaling-opentelemetry-collectors.md

---
title: Manage OpenTelemetry Collectors at scale with Ansible
linkTitle: OTel Collector with Ansible
date: 2024-03-12
author: '[Ishan Jain](https://github.com/ishanjainn) (Grafana)'
cSpell:ignore: ansible associated Ishan Jain
---

You can scale the deployment of
[OpenTelemetry Collector](/docs/collector/deployment/) across multiple Linux
hosts through [Ansible](https://www.ansible.com/), to function both as
[gateways](/docs/collector/deployment/gateway/) and
[agents](/docs/collector/deployment/agent/) within your observability
architecture. Using the OpenTelemetry Collector in this dual capacity enables a
robust collection and forwarding of metrics, traces, and logs to analysis and
visualization platforms.

We outline a strategy for deploying and managing the OpenTelemetry Collector's
scalable instances throughout your infrastructure using Ansible. In the
following example, we'll use [Grafana](https://grafana.com/) as the target
backend for metrics.

## Prerequisites

To follow this guide, make sure you meet the following requirements:

- Ansible Installed in your system
- Linux hosts along with SSH access to each of these Linux hosts

> - Prometheus for gathering metrics

Reviewer: Always found "diving in" a bit too idiomatic for ESL speakers. While more verbose, I think the former wording is clearer to all kind of audiences.

Agree on the bullet point edits — thanks for catching those!

> - Prometheus for gathering metrics

Reviewer 2: > Always found "diving in" a bit too idiomatic for ESL speakers.

Good point, thank you. How about:

- Before we begin, ...

   password. Make sure to pick a secure one.

For other installation methods and more detailed instructions, refer to the
[official Grafana documentation](https://grafana.com/docs/grafana/latest/#installing-grafana).

### Add Prometheus as a data source

1. In Grafana, navigate to **Connections** > **Data Sources**.
2. Click **Add data source** and select **Prometheus**.
3. In the settings, enter your Prometheus URL, for example,
   `http://<your_prometheus_host>`, along with any other necessary details.
4. Select **Save & Test**.

### Explore your metrics

1. Go to the **Explore** page
2. In the Query editor, select your data source and enter the following query

   ```PromQL
   100 - (avg by (cpu) (irate(system_cpu_time{state="idle"}[5m])) * 100)
   ```

   This query calculates the average percentage of CPU time not spent in the
   "idle" state, across each CPU core, over the last 5 minutes.

3. Explore other metrics and create dashboards to gain insights into your
   system's performance.

This guide makes it easier for you to set up the OpenTelemetry Collector on
several Linux machines with the help of Ansible and shows you how to see the

> data it collects in Grafana.

Reviewer 2: Hmm, not really. But given that @svrnm and @theletterf have given their thumbs up for the content, I'm willing to let this one pass.

> data it collects in Grafana.

Reviewer 3: let's leave it as is then, thanks @chalin

---
title: Manage OpenTelemetry Collectors at Scale with Ansible
linkTitle: OTel Collector with Ansible
date: 2024-03-12
author: '[Ishan Jain](https://github.com/ishanjainn) (Grafana)'
cSpell:ignore: ansible Ishan Jain
---

This guide is focused on scaling the
[OpenTelemetry Collector deployment](/docs/collector/deployment/) across various
Linux hosts by leveraging [Ansible](https://www.ansible.com/), to function both
as [gateways](/docs/collector/deployment/gateway/) and
[agents](/docs/collector/deployment/agent/) within your observability
architecture. Utilizing the OpenTelemetry Collector in this dual capacity
enables a robust collection and forwarding of metrics, traces, and logs to
analysis and visualization platforms.

> Here, we outline a strategy for deploying and managing the OpenTelemetry
> Collector's scalable instances throughout your infrastructure with Ansible,
> enhancing your overall monitoring strategy and data visualization capabilities
> in [Grafana](https://grafana.com/).

Reviewer: I think it should be less obvious why you chose Grafana when it could have been any other backend. Let's rephrase to something like this:

```suggestion
We outline a strategy for deploying and managing the OpenTelemetry
Collector's scalable instances throughout your infrastructure using Ansible.
In the following example, we'll use Grafana as the target backend.
```

## Before you begin

To follow this guide, ensure you have:

> - Ansible Installed in your system
> - Linux hosts along with SSH access to each of these Linux hosts.
> - Prometheus for gathering metrics

Reviewer: ```suggestion
- Ansible Installed in your system
- Linux hosts along with SSH access to each of these Linux hosts
- Prometheus for gathering metrics
```

## Install the Grafana Ansible collection

The
[OpenTelemetry Collector role](https://github.com/grafana/grafana-ansible-collection/tree/main/roles/opentelemetry_collector)
is provided through the
[Grafana Ansible collection](https://docs.ansible.com/ansible/latest/collections/grafana/grafana/)
as of the 3.0.0 release.

To install the Grafana Ansible collection, run this command:

```shell
ansible-galaxy collection install grafana.grafana
```

## Create an Ansible inventory file

> Next, you will set up your hosts and create an inventory file.

Reviewer: ```suggestion
Next, set up your hosts and create an inventory file.
```

1. Create your hosts and add public SSH keys to them.

   This example uses eight Linux hosts: two Ubuntu hosts, two CentOS hosts, two
   Fedora hosts, and two Debian hosts.

2. Create an Ansible inventory file.

   The Ansible inventory, which resides in a file named `inventory`, looks
   similar to this:

   ```ini
   10.0.0.1    # hostname = ubuntu-01
   10.0.0.2    # hostname = ubuntu-02
   10.0.0.3    # hostname = centos-01
   10.0.0.4    # hostname = centos-02
   10.0.0.5    # hostname = debian-01
   10.0.0.6    # hostname = debian-02
   10.0.0.7    # hostname = fedora-01
   10.0.0.8    # hostname = fedora-02
   ```

>    > **Note**: If you are copying the above file, remove the comments (#).

Reviewer: ```suggestion
   > **Note**: When copying the previous snippet, uncomment each host name.
```

Why?

3. Create an `ansible.cfg` file within the same directory as `inventory`, with
   the following values:

   ```cfg
   [defaults]
   inventory = inventory  # Path to the inventory file
   private_key_file = ~/.ssh/id_rsa   # Path to my private SSH Key
   remote_user=root
   ```

## Use the OpenTelemetry Collector Ansible role

> Next, you'll define an Ansible playbook to apply your chosen or created

Reviewer: ```suggestion
Next, define an Ansible playbook to apply your chosen or created
```

        otel_collector_exporters:
          prometheusremotewrite:
            endpoint: https://<prometheus-url>/api/prom/push
            headers:
              Authorization: 'Basic <base64-encoded-username:password>'

        otel_collector_service:
          pipelines:
            metrics:
              receivers: [hostmetrics]
              processors:
                [
                  resourcedetection,
                  transform/add_resource_attributes_as_metric_attributes,
                  batch,
                ]
              exporters: [prometheusremotewrite]
```

{{% alert title="Note" %}}

You'll need to adjust the configuration to match the specific telemetry you
intend to collect and where you plan to forward it. This configuration snippet
is a basic example designed for collecting host metrics and forwarded to
Prometheus.

{{% /alert %}}

The previous configuration would provision the OpenTelemetry Collector to
collect host metrics from the Linux host.

## Running the Ansible playbook

Deploy the OpenTelemetry Collector across your hosts by executing:

```sh
ansible-playbook deploy-opentelemetry.yml
```

> ## Visualizing Metrics in Grafana

Reviewer: ```suggestion
## Check your metrics in the backend
```

Once your OpenTelemetry Collector's start sending metrics to Prometheus, follow
these quick steps to visualize them in [Grafana](https://grafana.com/):

> ### Setup Grafana

Reviewer: ```suggestion
### Set up Grafana
```

1. **Install Docker**: Make sure Docker is installed on your system. If it's
   not, you can find the installation guide at the
   [official Docker website](https://docs.docker.com/get-docker/).

2. **Run Grafana Docker Container**: Start a Grafana server with this Docker
   command, which fetches the latest Grafana image:

   ```sh
   docker run -d -p 3000:3000 --name=grafana grafana/grafana
   ```

3. **Access Grafana**: Navigate to `http://localhost:3000` in your web browser.
   The default login details are `admin` for both the username and password.

4. **Change Password**: Upon your first login, you will be prompted to set a new
   password. Make sure to pick a secure one.

For other installation methods and more detailed instructions, refer to the

> [official Grafana documentation](https://grafana.com/docs/grafana/latest/installation/).

Reviewer: ```suggestion
official Grafana documentation.
```

> ### Add Prometheus as a Data Source

Reviewer: ```suggestion
### Add Prometheus as a data source
```

> 1. **Login to Grafana** and navigate to **Connections** > **Data Sources**.

Reviewer: ```suggestion
1. **Log in to Grafana** and navigate to **Connections** > **Data Sources**.
```

Reviewer: How do you access Grafana's login screen?

> 2. Click **Add data source**, and choose **Prometheus**.

Reviewer: ```suggestion
2. Click **Add data source** and select **Prometheus**.
```

> 3. In the settings, enter your Prometheus URL, for example,
>    `http://<your_prometheus_host>`, along with any other necessary details, and
>    then click **Save & Test**.

Reviewer: ```suggestion
3. In the settings, enter your Prometheus URL, for example,
   `http://<your_prometheus_host>`, along with any other necessary details.
4. Select **Save & Test**.
```

> ### Explore metrics

Reviewer: ```suggestion
### Explore your metrics
```

1. Go to the **Explore** page

> 2. In the Query editor, select your Prometheus data source and enter the below

Reviewer: ```suggestion
2. In the Query editor, select your data source and enter the following
```

   query

   ```PromQL
   100 - (avg by (cpu) (irate(system_cpu_time{state="idle"}[5m])) * 100)
   ```

   This query calculates the average percentage of CPU time not spent in the
   "idle" state, across each CPU core, over the last 5 minutes.

> 3. Play around with different metrics and start putting together your dashboards

Reviewer: How does this relate to scaling Collectors deployment through Ansible?

   to gain insights into your system's performance.

> This guide makes it easier for you to set up the OpenTelemetry Collector on
> several Linux machines with the help of Ansible and shows you how to see the
> data it collects in Grafana. You can adjust the settings for the OpenTelemetry
> Collector and design your Grafana dashboards to meet your own monitoring and
> tracking needs. This way, you get exactly the insights you want from your
> systems, making your job as a developer a bit easier.

Reviewer: ```suggestion
This guide makes it easier for you to set up the OpenTelemetry Collector on
several Linux machines with the help of Ansible and shows you how to see the
data it collects in Grafana.
```
